### PR TITLE
feat: SVM priority fee oracles and transaction submitters, more consistent tx submission

### PR DESCRIPTION
**Pull request by tkporter**
### Description

The fruition of https://www.notion.so/hyperlanexyz/Landing-Solana-Transactions-15a6d35200d680cbaf5eea5f9dbec1b1

Major highlights:
- Instead of using the max CUs, estimate CU usage and apply a 10% buffer
- Allow for priority fees to be configured. If not specified, defaults to a constant priority fee of 0. Supported options are:
  - a constant fee
  - the Helius priority fee API (https://docs.helius.dev/solana-apis/priority-fee-api)
- Allow for transaction submission to be configured. If not specified, defaults to the configured RPC url for the chain. Supported options are:
  - An arbitrary RPC URL, which will use priority fee auctions for fees
  - Jito, which will use a tip for fees
- In our own infra, we have logic to plumb through the Helius URL found at https://console.cloud.google.com/security/secret-manager/secret/mainnet3-rpc-endpoint-helius-solanamainnet/versions?project=abacus-labs-dev
- Signs txs with a finalized blockhash, as suggested by [How to Land Transactions on Solana](https://www.helius.dev/blog/how-to-land-transactions-on-solana)
- Partially implements `process_estimate_costs` to test that messages are processable. Does not yet actually enforce gas limits, because this will take some rollout effort. This is part of https://www.notion.so/hyperlanexyz/Alerting-on-critical-processing-issues-for-any-and-all-messages-15b6d35200d680a79d40fabc2e4e49da

### CODE REVIEW

File: rust/main/chains/hyperlane-sealevel/src/rpc/client.rs [hunk 1]
```diff
@@ -245,19 +320,174 @@ impl SealevelRpcClient {
...
             .map_err(ChainCommunicationError::from_other)?
-            .value
-            .return_data;
+            .value;
+
+        Ok(result)
+    }
+
+    /// Gets the estimated costs for a given instruction.
+    pub async fn get_estimated_costs_for_instruction(
```
Comment (tkporter) [inline rust/main/chains/hyperlane-sealevel/src/rpc/client.rs:345]: honestly not sure if these fns should live here, in provider.rs, or somewhere else - curious if anyone has strong takes

File: rust/main/helm/hyperlane-agent/templates/relayer-statefulset.yaml [hunk 2]
```diff
@@ -17,7 +17,9 @@ spec:
     metadata:
       annotations:
         checksum/configmap: {{ include (print $.Template.BasePath "/configmap.yaml") . | sha256sum }}
+        checksum/external-secret: {{ include (print $.Template.BasePath "/external-secret.yaml") . | sha256sum }}
```
Comment (tkporter) [inline rust/main/helm/hyperlane-agent/templates/relayer-statefulset.yaml:20]: drive-bys to make sure pods get updated if doing a release and there's a checksum change of the external secrets

File: typescript/infra/config/environments/mainnet3/agent.ts [hunk 3]
```diff
@@ -501,7 +550,7 @@ const releaseCandidate: RootAgentConfig = {
     rpcConsensusType: RpcConsensusType.Fallback,
     docker: {
       repo,
-      tag: '4cb2c9a-20241205-142854',
+      tag: 'b98678c-20241213-043725',
```
Comment (tkporter) [inline typescript/infra/config/environments/mainnet3/agent.ts:553]: you can see this running in RC atm https://solscan.io/account/G6fWKEBSrDtR8Uzx4kdVRRhrsHZ6zmGX1G2TYMmMgK6u - though might run out of funds. I'd expect some reverts as it's competing with the hyperlane relayer

check out the fees!

File: typescript/infra/src/config/agent/agent.ts [hunk 4]
```diff
@@ -85,8 +87,21 @@ export interface AgentContextConfig extends AgentEnvConfig {
...
+  priorityFeeOracleConfigGetter?: (
+    chain: ChainName,
+  ) => AgentSealevelPriorityFeeOracle;
+  transactionSubmitterConfigGetter?: (
+    chain: ChainName,
+  ) => AgentSealevelTransactionSubmitter;
+}
+
+// An ugly way to mark a URL as a the secret Helius URL when Helm templating
+export const HELIUS_SECRET_URL_MARKER = 'helius';
```
Comment (tkporter) [inline typescript/infra/src/config/agent/agent.ts:103]: intent here is to let us easily say in code "I want the helius RPC URL here" and have external secrets replace it

File: rust/main/chains/hyperlane-sealevel/src/mailbox.rs [hunk 5]
```diff
@@ -508,133 +463,40 @@ impl Mailbox for SealevelMailbox {
...
-        tracing::info!(?txn, "Created sealevel transaction to process message");
+        // Wait for the transaction to be confirmed.
+        self.rpc().wait_for_transaction_confirmation(&tx).await?;
 
-        let signature = self.send_and_confirm_transaction(&txn).await?;
-
-        tracing::info!(?txn, ?signature, "Sealevel transaction sent");
+        tracing::info!(?tx, ?signature, time_to_confirm=?send_instant.elapsed(), "Sealevel transaction confirmed");
 
+        // TODO: not sure if this actually checks if the transaction was executed / reverted?
```
Comment (tkporter) [inline rust/main/chains/hyperlane-sealevel/src/mailbox.rs:498]: was poking into this and I think we may need to change this - iiuc TxOutcome.executed is supposed to be false if the tx reverted, but I think `executed` will be true just if the tx lands

File: rust/main/config/mainnet_config.json [hunk 6]
```diff
@@ -785,7 +785,7 @@
       "index": {
         "from": 1,
         "mode": "sequence",
-        "chunk": 100
+        "chunk": 20
```
[the remark this answers is not in the document]
Comment (paulbalaji) [reply rust/main/config/mainnet_config.json:788]: could you also update the metadata in registry pls so future config generation respects the values 🙏

File: typescript/sdk/src/metadata/agentConfig.ts [hunk 7]
```diff
@@ -201,6 +255,13 @@ export const AgentChainMetadataSchema = ChainMetadataSchemaObject.merge(
       }
     }
 
+    // If the protocol type is Sealevel, require everything in AgentSealevelChainMetadataSchema
+    if (metadata.protocol === ProtocolType.Sealevel) {
+      if (!AgentSealevelChainMetadataSchema.safeParse(metadata).success) {
+        return false;
+      }
+    }
+
```
Comment (paulbalaji) [inline typescript/sdk/src/metadata/agentConfig.ts:264]: i can't see where these new params are in the agent config?

File: .changeset/long-llamas-fly.md [hunk 8]
```diff
@@ -1,5 +1,5 @@
 ---
-"@hyperlane-xyz/cli": patch
+'@hyperlane-xyz/cli': patch
 ---
 
 Suppress help on CLI failures
```
Comment (paulbalaji) [inline .changeset/long-llamas-fly.md:5]: need a new changeset for the SDK changes

File: rust/main/chains/hyperlane-sealevel/src/priority_fee.rs [hunk 9]
```diff
@@ -0,0 +1,225 @@
+#![allow(dead_code)]
```
Comment (ameten) [inline rust/main/chains/hyperlane-sealevel/src/priority_fee.rs:1]: We can remove it if we add annotations directly to fields which are not used:

```
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct JsonRpcResult<T> {
    #[allow(dead_code)]
    jsonrpc: String,
    #[allow(dead_code)]
    id: String,
    result: T,
}
```

File: rust/main/chains/hyperlane-sealevel/src/tx_submitter.rs [hunk 10]
```diff
@@ -0,0 +1,121 @@
+#![allow(dead_code)]
```
Comment (ameten) [inline rust/main/chains/hyperlane-sealevel/src/tx_submitter.rs:1]: We can remove it if we remove this unused function:

```
pub fn new_with_rpc_client(rpc_client: SealevelRpcClient) -> Self {
        Self { rpc_client }
    }
```

File: rust/main/chains/hyperlane-sealevel/src/mailbox.rs [hunk 11]
```diff
@@ -655,10 +517,30 @@ impl Mailbox for SealevelMailbox {
...
+        metadata: &[u8],
     ) -> ChainResult<TxCostEstimate> {
-        // TODO use correct data upon integrating IGP support
+        // Getting a process instruction in Sealevel is a pretty expensive operation
+        // that involves some view calls. Consider reusing the instruction with subsequent
+        // calls to `process` to avoid this cost.
+        let process_instruction = self.get_process_instruction(message, metadata).await?;
+
+        // The retuend costs are unused at the moment - we simply want to perform a simulation to
+        // determine if the message will revert or not.
```
Comment (ameten) [inline rust/main/chains/hyperlane-sealevel/src/mailbox.rs:529]:
```suggestion
        // The retuned costs are unused at the moment - we simply want to perform a simulation to
        // determine if the message will revert or not.
```

**Review by ameten — APPROVED** [review]

File: typescript/sdk/src/metadata/agentConfig.ts [hunk 12]
```diff
@@ -201,6 +255,13 @@ export const AgentChainMetadataSchema = ChainMetadataSchemaObject.merge(
       }
     }
 
+    // If the protocol type is Sealevel, require everything in AgentSealevelChainMetadataSchema
+    if (metadata.protocol === ProtocolType.Sealevel) {
+      if (!AgentSealevelChainMetadataSchema.safeParse(metadata).success) {
+        return false;
```
Comment (jmrossy) [inline typescript/sdk/src/metadata/agentConfig.ts:261]: nit: `return AgentSealevelChainMetadataSchema.safeParse(metadata).success`

File: rust/main/chains/hyperlane-sealevel/src/rpc/client.rs [hunk 13]
```diff
@@ -245,19 +320,174 @@ impl SealevelRpcClient {
...
+        let simulation_compute_units: u32 = simulation_result
+            .units_consumed
+            .and_then(|units| if units > 0 { Some(units) } else { None })
+            .ok_or_else(|| {
+                ChainCommunicationError::from_other_str(
+                    "Empty or zero compute units returned in simulation result",
+                )
+            })?
+            .try_into()
+            .map_err(ChainCommunicationError::from_other)?;
```
Comment (daniel-savu) [inline rust/main/chains/hyperlane-sealevel/src/rpc/client.rs:388]: nit
```suggestion
        let simulation_compute_units: u32 = simulation_result.units_consumed.unwrap_or_default();
        if simulation_compute_units.is_zero() {
            return Err(ChainCommunicationError::from_other_str(
                "Empty or zero compute units returned in simulation result",
            ));
        }
```

File: rust/main/chains/hyperlane-sealevel/src/mailbox.rs [hunk 14]
```diff
@@ -508,133 +463,40 @@ impl Mailbox for SealevelMailbox {
         metadata: &[u8],
         _tx_gas_limit: Option<U256>,
```
Comment (daniel-savu) [inline rust/main/chains/hyperlane-sealevel/src/mailbox.rs:464]: it's a bit confusing to have this function called with `Some(0)` value (from the `process_estimate_costs` result) to then ignore it. I think we'd respect the trait interface better if we returned `gas_limit: None` in `process_estimate_costs`, called `unwrap_or_default` on it when doing the igp check, and feeding this argument into `build_estimated_tx_for_instruction`

File: rust/main/chains/hyperlane-sealevel/src/mailbox.rs [hunk 15]
```diff
@@ -508,133 +463,40 @@ impl Mailbox for SealevelMailbox {
...
+        let send_instant = std::time::Instant::now();
 
-        tracing::info!(?txn, "Created sealevel transaction to process message");
+        // Wait for the transaction to be confirmed.
+        self.rpc().wait_for_transaction_confirmation(&tx).await?;
 
-        let signature = self.send_and_confirm_transaction(&txn).await?;
-
-        tracing::info!(?txn, ?signature, "Sealevel transaction sent");
+        tracing::info!(?tx, ?signature, time_to_confirm=?send_instant.elapsed(), "Sealevel transaction confirmed");
```
Comment (daniel-savu) [inline rust/main/chains/hyperlane-sealevel/src/mailbox.rs:496]: isn't the `send_instant.elapsed()` more of a result of the commitment level used? Especially since there's also a confirm check below, suggesting we don't yet have confirmation here

File: rust/main/chains/hyperlane-sealevel/src/rpc/client.rs [hunk 16]
```diff
@@ -199,16 +216,74 @@ impl SealevelRpcClient {
             .map_err(ChainCommunicationError::from_other)
     }
 
-    pub async fn send_and_confirm_transaction(
+    pub async fn send_transaction(
         &self,
         transaction: &Transaction,
+        skip_preflight: bool,
```
Comment (daniel-savu) [inline rust/main/chains/hyperlane-sealevel/src/rpc/client.rs:222]: when would we set this to `false`?

File: rust/main/chains/hyperlane-sealevel/src/rpc/client.rs [hunk 17]
```diff
@@ -245,19 +320,174 @@ impl SealevelRpcClient {
...
+        instruction: Instruction,
+        payer: &Keypair,
+        tx_submitter: &dyn TransactionSubmitter,
+        sign: bool,
+    ) -> ChainResult<Transaction> {
+        let instructions = vec![
+            // Set the compute unit limit.
+            ComputeBudgetInstruction::set_compute_unit_limit(compute_unit_limit),
+            // Set the priority fee / tip
+            tx_submitter.get_priority_fee_instruction(
```
Comment (daniel-savu) [inline rust/main/chains/hyperlane-sealevel/src/rpc/client.rs:459]: nice, so this means you can't set both a priority fee and use jito, which would be ineffective 🔥

File: rust/main/chains/hyperlane-sealevel/src/tx_submitter.rs [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+        ComputeBudgetInstruction::set_compute_unit_price(compute_unit_price_micro_lamports)
+    }
+
+    async fn send_transaction(
+        &self,
+        transaction: &Transaction,
+        skip_preflight: bool,
+    ) -> ChainResult<Signature> {
+        self.rpc_client
+            .send_transaction(transaction, skip_preflight)
```
Comment (daniel-savu) [inline rust/main/chains/hyperlane-sealevel/src/tx_submitter.rs:66]: when you say this approach uses priority fee auctions in the PR description, is that implicit to this call?

**Review by daniel-savu — APPROVED** [review]

File: rust/main/config/mainnet_config.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (tkporter) [reply rust/main/config/mainnet_config.json:788]: https://github.com/hyperlane-xyz/hyperlane-registry/pull/454

File: typescript/sdk/src/metadata/agentConfig.ts (hunk 7 above)
Comment (tkporter) [reply typescript/sdk/src/metadata/agentConfig.ts:265]: it's this merge https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/3a62b2df86efe6b6d3186266e6f689eb3020317d/typescript/sdk/src/metadata/agentConfig.ts#L212

File: rust/main/chains/hyperlane-sealevel/src/priority_fee.rs (hunk 9 above)
Comment (tkporter) [reply rust/main/chains/hyperlane-sealevel/src/priority_fee.rs:1]: ty, forgot to remove this

File: typescript/sdk/src/metadata/agentConfig.ts (hunk 12 above)
Comment (tkporter) [reply typescript/sdk/src/metadata/agentConfig.ts:262]: think it may be better keep as-is so that if anyone adds more checks after this line in the future, we don't return true earlier than we actually want to. Not sure it'd be obvious to someone of the sharp edge there. I assume that was the rationale for the cosmos parsing a little above this

File: rust/main/chains/hyperlane-sealevel/src/priority_fee.rs [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+            serde_json::json!({
+                "recommended": true,
+                "transactionEncoding": "base58",
+            })
+        } else {
+            serde_json::json!({
+                "priorityLevel": self.config.fee_level,
+                "transactionEncoding": "base58",
+            })
+        }
```
[the remark this answers is not in the document]
Comment (tkporter) [reply rust/main/chains/hyperlane-sealevel/src/priority_fee.rs:67]: much nicer, was spooked by the serde_json types lol

File: rust/main/chains/hyperlane-sealevel/src/mailbox.rs (hunk 14 above)
Comment (tkporter) [reply rust/main/chains/hyperlane-sealevel/src/mailbox.rs:464]: agree it's weird, though think we should maybe just leave this to a later PR

File: rust/main/chains/hyperlane-sealevel/src/mailbox.rs (hunk 15 above)
Comment (tkporter) [reply rust/main/chains/hyperlane-sealevel/src/mailbox.rs:496]: hmm yes, I'm adding a comment about this but gonna leave logic as is atm. I think it's probably still useful

File: rust/main/chains/hyperlane-sealevel/src/rpc/client.rs (hunk 16 above)
Comment (tkporter) [reply rust/main/chains/hyperlane-sealevel/src/rpc/client.rs:222]: hmm good point - despite us not setting it to false atm I think it's a little nice to have in the interface as a way to indicate that we're doing something a little non-standard. But don't feel strongly. Gonna leave as is and we can change later if we want

File: rust/main/chains/hyperlane-sealevel/src/rpc/client.rs (hunk 17 above)
Comment (tkporter) [reply rust/main/chains/hyperlane-sealevel/src/rpc/client.rs:461]: yep! funnily enough Jito has some confusing docs where if you choose to use the non-MEV protected endpoint, they want you to use both priority fees and the jito tips. But we just won't support that for now to keep a nicer interface

File: rust/main/chains/hyperlane-sealevel/src/tx_submitter.rs (hunk 18 above)
Comment (tkporter) [reply rust/main/chains/hyperlane-sealevel/src/tx_submitter.rs:66]: yep exactly
